### ISSUE TITLE
Code review needed

### ISSUE THREAD

**Issue by JamesSample** [NIVANorge/aquamonitR#1]
@DagHjermann, @vivigirardin, @BennoDillinger, @RaoulWolf

Dear NIVA R gurus,

I'm not sure how useful this repository will be, but @roarbra has been improving/extending the [Aquamonitor-Python](https://github.com/NIVANorge/Aquamonitor-Python) API recently and I thought it might be useful to have a native R version too. Functionality is limited at present, but you can e.g. query all the water chemistry data for a particular project directly as an R dataframe, which I think will be useful for some people. It's also not dependent on a direct/internal database connection, so you could use it e.g. for Shiny apps deployed externally.

_However_, I don't write much R code, so everything here could do with a review/tidy from someone that actually knows what they're doing. In particular, I always find classes in R confusing (S3, S4, R6, "reference classes"... argh!), so any feedback appreciated. I've used a reference class here because it looks vaguely like Roar's Python implementation, but alternative approaches welcome.

No hurry whatsoever, and if this doesn't look useful just ignore it for now. Once the Python version is a bit more comprehensive I'll return to this, but that probably won't be for a while.

Thank you!

And a very merry Christmas to you all :-)

**Comment by RaoulWolf** [comment]
Hi James et al,

and first of all happy new year! 🎆  I hope everyone had *some* time off and everyone is happy and healthy 😃 

I'll take a quick look at the code later this week; first there's an unhealthy pile of e-mails that requires my attention...

If you want to turn this into an (internal) R package, I have a bit of experience, e.g., from my [{chemspiderapi}](https://github.com/NIVANorge/chemspiderapi) package. {aquamonitR} would likely not need *all* the fancy stuff, but I can quickly set things up once we know if this is desired.

**Comment by DagHjermann** [comment]
Happy new year to the two of you also! All happy and healthy here. Even managed a very nice trip to my wife's family cabin in the mountains this weekend in spite of some difficulties! *)

Very cool initiative! I wasn't able to log in, though. Is it supposed to work on my computer (not inside Jupyterhub), user = NIVA initials (3 capital letters) and password = ordinary NIVA password? 

(Also the script I made in Jupyterhub for getting data from Aquamonitor doesn't work again, sigh .... don't know if there is any connection)

*) the suspension of our car is broken, so we borrowed the car of my father-in-law, and then the exhaust pipe broke off just behind the engine so we had to use a strap between two doors to avoid losing it completely - but we got there in the end!

**Comment by JamesSample** [comment]
Happy New Year!

Thanks @RaoulWolf for offering to look at the code!

Sorry to hear it didn't work for you @DagHjermann. Did you remember to set the `host` after sourcing the file, as in the [Quickstart](https://github.com/NIVANorge/aquamonitR#quick-start) example?

``` r
# The new functionality requires using the AM dev/test site
host <- "https://test-aquamonitor.niva.no/"
``` 

The username and password you're using sound correct - it should be the same credentials you use to login to the [Aquamonitor user interface](http://www.aquamonitor.no/Intern/). I've only tested it on the JupyterHub so far but, other than being a Linux container, there shouldn't be anything special about running it there. I think it should work from anywhere.

What error are you seeing? Is it just `Login failed. Please check your username and password.` ?

If you clone this repo to your JupyterHub and run the example notebook (or the Quickstart code via R-Studio), does it work for you then?

Your original JupyterHub script is probably not working because Roar made some changes to the API backend around the time you gave your TechTalk last year. The hacky solution we were using previously therefore needs adapting to match Roar's changes. I haven't looked at this yet (because I was hoping to use Aquamonitor-Python/AquamonitR instead, if possible), but if you need these scripts now I can probably update the hacky method as a temporary fix?

Thanks! :-)

P.S. Your trip to the cabin sounds exciting!

**Comment by DagHjermann** [comment]
Host not resolved, see below. Sorry I didn't write that at once.
![bilde](https://user-images.githubusercontent.com/2622685/103644642-fd46f600-4f56-11eb-8852-ef80c329b54f.png)

I will try in Jupyterhub! (and I don't need to use my original Jupyterhub solution now, I'll rather use AquamonitR istead!)

all the best, Dag

**Comment by DagHjermann** [comment]
Ahem, I tried again on my computer and now I got " Login failed. Please check your username and password." - meaning that the response was NoUser. I have an "ø" in my password, could that be the problem? (I can change password before I quit today and test again tomorrow.)

Tried in Jupyterhub, there I got a 404 error:
![bilde](https://user-images.githubusercontent.com/2622685/103649263-78f87100-4f5e-11eb-9cd4-d85ce79c4b89.png)

**Comment by roarbra** [comment]
Happy New Year, everybody!

This is marvelous work @JamesSample .
There are one thing I would like to comment on though. In AquaMonitor we have this machinery to normalize parameter names and units. When you're asking for the table water_chemistry_input. You're not getting that data, but the unhandled data.
For R usage I think we should focus on the table "water_chemistry_output". I've worked on it and belive that it's ready. Here you can see the outcome of a single measurement:

```
{
  '$type': 'Niva.AquaMonitor.Cargo.ValueCargo, Niva.AquaMonitor.Cargo',
  'Parameter': {
    'Id': 301,
    'Name': 'Temperatur',
    'Unit': 'C'},
  'Sample': {
    '$type': 'Niva.AquaMonitor.Cargo.WaterSampleCargo, Niva.AquaMonitor.Cargo',
    'Id': 11354644,
    'Depth1': 0.0,
    'Depth2': 0.0,
    'Method': {},
    'SampleDate': '2019-09-30T00:00:00Z',
    'Station': {
      'Id': 26405,
      'Project': {
        'Id': 12433,
        'Name': 'Nasjonal Innsjøundersøkelse 2019'},
      'Type': {},
      'Code': '834-2-5',
      'Name': 'Vonevatn'}},
  'Value': 5.6}
```

In addition to normalizing names and units, you will not get the none approved values. And therefore not the field Approved.

The usage of the table water_chemistry_input would be more for quality check and maintenance.

**Comment by roarbra** [comment]
> Ahem, I tried again on my computer and now I got " Login failed. Please check your username and password." - meaning that the response was NoUser. I have an "ø" in my password, could that be the problem? (I can change password before I quit today and test again tomorrow.)
> 

We've had some difficulties with strange characters, but I think they're fixed.

**Comment by JamesSample** [comment]
@DagHjermann Can you try again on the JupyterHub please, but make sure you set the `host` **_after_** you've sourced the aquamonitR.R file (see the file `shared/DHJ/aquamonitR/test.R` as an example). The host is initially set when you source aquamonitR.R [here](https://github.com/NIVANorge/aquamonitR/blob/9238cd38035cabc618456465e2d357daa1b36d88/aquamonitR.R#L8), but you need to switch to the dev/test URL for the new code to work. At present, you're setting `host` correctly first, but then setting it back to the original when you source the file.

If I run `test.R` from your folder on `shared`, I can login OK with my credentials, so strange if it doesn't work for you too.

@roarbra Thanks for the tip - I'll look into this. If I understand you correctly, this will also fix the issue [here](https://github.com/NIVANorge/Aquamonitor-Python/issues/2) with Aquamonitor-Python?

**Comment by JamesSample** [comment]
@roarbra I've now switched to using the `water_chemistry_output` table, as suggested.

Relevant commit: b5b18c33e024f587d7fae1f9005ff6e74aa9bf0b

**Comment by DagHjermann** [comment]
@JamesSample Argh, that was a stupid mistake I did there! Now it works!

Or rather, logging in, reading projects and stations works. `get_project_chemistry` fails due to a small thing: It tries to change the names of the variables `Sample.Station.Project._Id` and `Sample.Station.Project._Name`, but when I run it, those variables come out without the underscores (i.e., I get `Sample.Station.Project.Id` and `Sample.Station.Project.Name`).  

It still doesn`t work on my Windows computer though. I think it is an encoding problem in R, or perhaps RStudio specifically (when using RStudio, as the `getPass`function calls `rstudioapi::askForPassword`). I'll try to change password to somethong with "normal" letters....

**Comment by JamesSample** [comment]
@DagHjermann Good to know it's working on the Hub, at least - thanks Dag! I think the issue with the column names is caused by Roar recently modifying something in the backend code. This is hopefully fixed by my latest commit (from an hour ago), which should also handle the conversion to standardised parameter names and units.

I'll install and test on Windows when I get chance. It must be some character escaping or unicode problem, as you say. Thanks for testing!

**Comment by roarbra** [comment]
> @DagHjermann Good to know it's working on the Hub, at least - thanks Dag! I think the issue with the column names is caused by Roar recently modifying something in the backend code. This is hopefully fixed by my latest commit (from an hour ago), which should also handle the conversion to standardised parameter names and units.
> 
You're absolutely right about that @JamesSample . Let me now wherever you see a field name with underscore. It's easy to fix.

**Comment by JamesSample** [comment]
Just tested on Windows and everything works OK for me, but there seems to be a bug with how R-Studio for Windows handles special characters. 

Using getPass via the old-style RGui handles special characters correctly, so I think this would work for you, @DagHjermann. However, as Dag mentions above, using getPass via R-Studio calls `rstudioapi::askForPassword`, which doesn't interpret special characters correctly on Windows. For example, with the following code:

``` r
library("getPass")
getPass("Type some special characters: ")
```

If I enter `øæå` via R-Studio on the JupyterHub (Linux), this returns `øæå`, which is correct. Doing the same using R-Studio for Windows returns `Ã¸Ã¦Ã¥`, which is wrong and leads to a login error.

I've not used R-Studio on Windows much (I just installed it now for testing). Maybe there are some configuration options that would fix this? Any suggestions @DagHjermann @RaoulWolf ?

Thanks!

(I'll leave this issue open for now pending code review).

**Comment by vivigirardin** [comment]
Hello and happy new year! Sorry about the late reply, but it is in my Latin blood...
 I run the code on Windows, and it works great!

**Comment by vivigirardin** [comment]
> If I enter `øæå` via R-Studio on the JupyterHub (Linux), this returns `øæå`, which is correct. Doing the same using R-Studio for Windows returns `Ã¸Ã¦Ã¥`, which is wrong and leads to a login error.
> 
I had the same issue with my Shiny app, which I managed to fix using  `encoding = "windows-1252"`

**Comment by RaoulWolf** [comment]
Should we then consider to automatically detect the OS of the user and change the encoding if necessary?

To which functions would this apply? I assume the {getPass} functions as well as the {HTTR} functions would be encoding sensitive, but are there any others?

Sorry I haven't had the time yet to check the code, but pinky promise I'll start with it today 😄

**Comment by DagHjermann** [comment]
Now it works both on Windows (after I changed my password) and on the hub. This is **super!** (I hope it sometime can be exdended to biota chemistry!) 

Just one small thing: remove `approved = TRUE` in the example in `README.md`.

Regarding the code review:   
> In particular, I always find classes in R confusing (S3, S4, R6, "reference classes"... argh!), so any feedback appreciated.   

Me too! To the extent that I never use them. So I'm not very helpful in this case! From a quick look, the rest of the code looks fine.

**Comment by JamesSample** [comment]
Thanks @vivigirardin! Where/how did you set the encoding?

@RaoulWolf Detecting the OS might be the solution, but I'd like to avoid doing this explicitly if possible. I'm trying to find a global setting to force R/R-Studio to interpret all strings as UTF-8, but no luck so far. Text encodings are always a nightmare!

@DagHjermann Thanks - I've updated the Quickstart example. I think @roarbra 's main `query` class (which I've reproduced here) probably already supports biota/ecology etc., I just haven't written a convenience function for it yet (like `get_project_chemistry`). [Aquamonitor-Python](https://github.com/NIVANorge/Aquamonitor-Python) has lots of examples beyond water chemistry, but I think some of them might need updating/tweaking to reflect the latest updates. I need to check with @roarbra regarding how best to access the different data types, but I think it should be fairly easy to create additional functions like `get_project_chemistry` that all use the same `query` class in the background.

**Comment by RaoulWolf** [comment]
The code on the `README` works well for me as well! I'll go ahead and take a closer look at the code now. 

I'll open a (some?) new issues if there's anything specific popping up.

**Comment by JamesSample** [comment]
I think I've fixed the issue with special characters in R-Studio on Windows by explicitly specifying the input encoding using e.g.

    Encoding(username) <- "UTF-8"

(the encoding is `unknown` by default). See ce234c5fe9acaee1e02f4aae90741a1a531a2eac for details.

I _think_ this will work on both Windows and Linux, because R-Studio can then convert the encoding automatically as necessary. This has no effect on R-Studio for Linux (which assumes UTF-8 by default), but does make a difference on Windows (which, by default, interprets the UTF-8 input as `latin1`/`windows-1252`, and therefore gets it wrong).

@RaoulWolf Great - thanks very much! Feel free to open more issues/pull requests as necessary :-)

**Comment by vivigirardin** [comment]
> 
> 
> I think I've fixed the issue with special characters in R-Studio on Windows by explicitly specifying the input encoding using e.g.
> 
> ```
> Encoding(username) <- "UTF-8"
> ```
I guess you already found a solution @JamesSample . 
Anyway, I don't use the NIVA username/password (haven't got there yet). I store the password in Renviron. I was just offering a hint :)
```
[      conn <-
          DBI::dbConnect(
            
            odbc::odbc(),
            'LW_Prod',
            UID = Sys.getenv('LABWAREUSER'), #- Username and password are stored in Renviron.
            PWD = Sys.getenv('LABWAREPASS'),
            encoding = "windows-1252"
            
          )
```

**Comment by JamesSample** [comment]
Closed by 926abba475da9d88c346f106f08d1d37cc4c353a.
